Richardson Lucy Parallelization V2

avalluvan opened the pull request:
Based on feedback that I received on version 1 of RL parallelization, I have incorporated a new setup.

**RichardsonLucy.py**

- Introduction of a new object `comm` that handles all MPI communication if a MPI descriptor is passed as an argument during initialization
- No changes to existing structure. Features such as acceleration parameter, background normalization, etc. re-use existing code
- 1 synchronization barrier at initialization (summed exposure map) and 3 barriers at every iteration (end of Estep - expectation, end of Mstep - delta_model, and postprocessing broadcast of updated model) to support message passing. 
- Introduction of "slice" variables to work with the partial vectors that are generated at each node
- Final output is exactly similar to earlier
- Works in both serial and parallel modes
- Serial mode (and single node parallel mode, which is effectively a serial code) can be accessed via a jupyter notebook (including the existing tutorial). For parallel mode, use RLparallelscript.py

**DataInterfaceWithParallelSupport.py**

- An iterative update for the data interface
- Instead of passing the data directly, one needs to pass the event, bg and image response filenames.
- Also pass the `comm` object
- Limitation: I am working towards expanding this to handling the FullDetectorResponse file. The NUMROWS and NUMCOLS are currently fixed but will use FullDetectorResponse.npix in the future
- The object `dataset` returned by this new module works exactly the same way as the DataInterfaceDC2 module. Pass it to image_deconvolution through `ImageDeconvolution.set_dataset([dataset])`
- Limitation: Many ad-hoc fixes for sliced data formats. Could be fixed with specific features from `histpy.Histogram` if they exist. Multiple instances of object reconstruction was required. 
- Ideally, two response files should be stores (the usual and the transposed version) to maximize speed upgrades from column major data reads. 
- Functions like calc_likelihood, calc_expectation, etc. remain the same except minor changes from "full" variables to "sliced" variables to accommodate each node's smaller execution size.

**RLparallelscript.py**

- A completely refreshed version of RLparallelscript.py
- Invoked as ```mpiexec -n <number of processes> python RLparallelscript.py```

avalluvan commented:
Currently, DC2 (existing) and Parallel (new) Data Interfaces can be used interchangeably for serial code. They produce the same output. However, the latter must be used for parallel code.

israelmcmc commented:
Thanks @avalluvan! I think it's a great improvement with respect to V1. I still need to look at the code in detail, but I read your description and checked the files changed. A few first impressions:

* About these two limitations:
> Limitation: I am working towards expanding this to handling the FullDetectorResponse file. The NUMROWS and NUMCOLS are currently fixed but will use FullDetectorResponse.npix in the future
> Limitation: Many ad-hoc fixes for sliced data formats. Could be fixed with specific features from histpy.Histogram if they exist. Multiple instances of object reconstruction was required.

It’s good that you open the PR so we can start the review, but I’d wait for these two limitations to be resolved before merging.

* Can you please clarify why you needed to modify `RichardsonLucy.py`? In the example I shared on #237 I was able to use MPI only by changing the data interface (all barriers were contained in the data interface of that class). I'd prefer to modify as few parts of the code as possible unless really necessary.

* In a similar spirit, I see that many other file changes seem unrelated or not needed for the RL parallelization, e.g., notebooks, config files, your personal comments in code, and probably some features related to the new response handling. Can you please keep only the changes needed for the features in this PR? That will facilitate the review. I'd be happy to help if you have questions on how to deal with git to achieve this.

avalluvan commented:
On point 1, I have updated the code to migrate most parallelization features to `dataIFWithParallelSupport.py`. There are three instances where the `RichardsonLucy` class needs to know which node is running it (serial mode/MASTER node or any other node), all of them pertaining to how results are processed. `register_results()` can, in principle, be stored in each node, however, if all the nodes were to save it to disk then that may cause unnecessary R/W slowdowns. Instead of passing the `comm` object to the `RichardsonLucy` class, do you suggest adding a flag variable? If yes, where should I add that - `image_deconvolution.py` or `RichardsonLucy.py`

I have resolved the merge conflicts in the response handling code. I had added a few comments to parts of the imaging code that took me a while to figure out for easier reading. Do you want me to remove those? The tutorial notebooks were probably modified greatly and I would not want to commit those changes to the develop branch.

Do you think we should wait till the dataIF code is modified for DC3 and handles FullDetectorResponse objects properly? The current pull request adds a feature for parallel execution on top of the existing DC2 imaging codes, and I think could be merged as an iterative update.

hiyoneda commented on cosipy/image_deconvolution/RichardsonLucy.py:
```diff
@@ -172,15 +193,15 @@ def post_processing(self):
-        
-        # update expectation_list
-        self.expectation_list = self.calc_expectation_list(self.model, dict_bkg_norm = self.dict_bkg_norm)
-        logger.debug("The expected count histograms were updated with the new model map.")
```
Can you keep these lines? Otherwise, the likelihood is evaluated with the old model.

hiyoneda commented on cosipy/image_deconvolution/RichardsonLucy.py:
```diff
@@ -112,9 +122,14 @@ def pre_processing(self):
+        logger.info("The expected count histograms were calculated with the initial model map.")
+
+        # At the end of this function, all processes should have a complete `self.expectation_list`
+        # to proceed to the Mstep function
```
You can remove these lines if they are performed in post_processing.

hiyoneda commented on cosipy/image_deconvolution/RichardsonLucy.py:
```diff
@@ -93,10 +107,6 @@ def initialization(self):
-        # expected count histograms
-        self.expectation_list = self.calc_expectation_list(model = self.initial_model, dict_bkg_norm = self.dict_bkg_norm)
-        logger.info("The expected count histograms were calculated with the initial model map.")
-
```
It is possible to keep these lines? To use the updated model for the likelihood calculation, I wanted to perform the expected count calculation at the post-processing and the initialization step and skip it in the Estep.

hiyoneda commented on cosipy/image_deconvolution/RichardsonLucy.py:
```diff
@@ -66,16 +70,26 @@ def __init__(self, initial_model, dataset, mask, parameter):
             else:
                 os.makedirs(self.save_results_directory)
 
```
I understand that RL needs to know if it is performed on the master node and needs this kind of parameter. I would suggest preparing two parameters alternatively, something like

- self.parallel_computation = True / False
- self.master_node = True / False

I want to prepare a parameter that explicitly tells if the computation is in parallel or not. I will add some suggestions regarding these changes at other lines.

hiyoneda commented on cosipy/image_deconvolution/RichardsonLucy.py:
```diff
@@ -193,21 +214,23 @@ def register_result(self):
+            # show intermediate results
+            logger.info(f'  alpha: {this_result["alpha"]}')
+            logger.info(f'  background_normalization: {this_result["background_normalization"]}')
+            logger.info(f'  loglikelihood: {this_result["loglikelihood"]}')
```
If you agree to add self.parallel_computation and self.master_node, I want to add the following lines at the beginning of this class function;
```
if self.parallel_computation == True and self.master_node == False:
    pass
```
I think that it is easier to understand what happens in the child nodes.

avalluvan replied:
Changes accepted.

hiyoneda commented on cosipy/image_deconvolution/RichardsonLucy.py:
```diff
@@ -225,38 +249,41 @@ def finalization(self):
         """
         finalization after running the image deconvolution
         """
```
The same as before. Can we add these lines at the beginning instead?
```
if self.parallel_computation == True and self.master_node == False:
    pass
```

avalluvan replied:
Changes accepted.

hiyoneda commented on cosipy/image_deconvolution/allskyimage.py:
```diff
@@ -151,9 +151,11 @@ def set_values_from_parameters(self, parameter):
             unit = u.Unit(algorithm_parameter['unit'])
             for idx, value in enumerate(algorithm_parameter['value']):
-                self[:,idx] = value * unit
+                self[:,idx] = value * unit      # TODO: Why is this indexed as [:, idx] instead of [idx, :]?
```
> Why is this indexed as [:, idx] instead of [idx, :]?

It is because the order of axes is defined as (lb, Ei). It follows the axes order of the full detector response.

hiyoneda reviewed (commented):
Thanks, @avalluvan. I added some comments on your RL codes directly. 

By the way, I noticed that you changed some classes which are probably not related to the RL parallelization itself. For example, FullDetectorResponse, SpacecraftFile, PointSourceResponse. I am concerning that reviewing these different issues simultaneously may cause mistakes easily. So, is it possible to separate them from this PR? Then, we can review this PR more easily.

avalluvan reviewed (commented):
Reviewed all changes. All files except point_source_injector.ipynb are intact.

avalluvan commented:
It looks like the unit tests are failing because `mpi4py` is not part of the package. How do I rectify that?

israelmcmc commented:
> It looks like the unit tests are failing because `mpi4py` is n

israelmcmc commented:
Thanks @avalluvan . I haven't checked all of this yet, but about this:

> It looks like the unit tests are failing because `mpi4py` is not part of the package. How do I rectify that?
In principle, you should update the requirements here: https://github.com/cositools/cosipy/blob/c87577d35a213f7a364e6cd2d428f7c46668ad09/setup.py#L25

However, mpi4py is a special case, because it needs to have the backend MPI installed, which I don't think you can do with pip (I used conda). One option is
- Since most users won't use mpi4py, we can make it an optional package. If it's not installed, then a try/except clause will catch it during the imports. The code can only run in series in that case, which is fine. 
- The above will make the test pass, but it won't actually test the parallelization. We should probably have a custom workflow where MPI is installed to test this. It's OK with me if we leave this for future work.

israelmcmc commented:
After a long delay (my apologies again!) I finally went through this PR in detail. I still think it was a great improvement from V1. Putting most of the MPI code in `DataIF_Parallel` plus the script really simplified things. 

There were still a few things I thought could be improved before merging, and in the interest of not delaying this further, I opened a [PR to your PR](https://github.com/avalluvan/cosipy/pull/2). Here are the changes:
- I synced your branch with the current develop and resolve all conflicts. Note that `DataIF_parallel` needed a new parameter `bkg_norm_label` in order to work with the latest changes.
- I reverted all changes in `RichardsonLucy`. In order to address the issue with `register_results()` --which should only run in the master node, as you mentioned above— I instead created a new class called `ParallelImageDeconvolution`, which derives from `ImageDeconvolution` and overload the `_finalize` method to only call the `DeconvolutionAlgorithmBase` implementation in the master node. The script knows whether it's running in parallel or series, and instantiate the correct `ImageDeconvolution` class. In addition to avoid messing around with `RichardsonLucy`, I think this will allow to run other algorithms in parallel without extra work (I haven't test it though).
-  The `mpi4py` imports are now conditional. Installing this package will be optional.
- I removed the hard coded values for the inputs files and the number of pixels. Now you can run the script as e.g.

```
 mpiexec -n 2 python RLparallelscript.py --config imagedeconvolution_parfile_gal_511keV.yml --data 511keV_dc2_galactic_event.hdf5 --bkg 511keV_dc2_galactic_bkg.hdf5 --response psr_gal_511_DC2.h5
```

This runs for me pretty well and gives me the same results with/without mpiexec.

Side note: Interestingly, running it with `mpiexec -n 2` was about 3-4 times faster than not using `mpiexec`. I was expecting 2x. Maybe there are other dependencies that benefit from mpi, idk. It would be good to understand this.

@avalluvan If you agree with the changes and don't see any issues, please merge my PR to your PR, and I'll be happy to finally merge this! It'll be more important to have it in cosipy now that we're getting closer to the unbinned analysis.

@hiyoneda Please take a look, but I think after my changes, all remaining changes to the existing code in the imaging class are very minor (mostly fixing typos and adding type hints). Most of the new code has been contained in new classes, which is beauty of interfaces: they can't harm the main code!

For references, this is the diff between my branch and develop (how this PR would look like if Anaya merges my changes)
https://github.com/cositools/cosipy/compare/develop...israelmcmc:PR274-Israel

avalluvan commented:
Hi Israel. I agree with most of your modifications. I have two specific questions:

- How can we future-proof the sections where we load the response/response-transpose (dataIF_Parallel.py, `_MPI_load_data()`)? It is currently designed for DC2 with overflow and underflow bins and contains hardcoded adjustments, e.g. 
    - `R = dataset[1:-1, 1:-1, 1:-1, 1:-1, start_col+1:end_col+1]`
    - `nrows = dataset.shape[0] - 2`
- Do we need a plotting routine at the end of RLparallelscript.py? If no, we could remove that section (L133-L153) and the associated library imports (matplotlib and mhealpy). Alternately, can we control it using a boolean flag (plot_result = True) in the config file?

The `mpiexec` speed gains are indeed interesting. I am sure h5py benefits from MPI - If `Histogram.open()` uses h5py in the backend that could be a reason.

israelmcmc commented:
@avalluvan 

* The optional plotting makes a lot of sense. I added it mostly for debugging purposes. I added a `--plot` flag to make it optional, including the imports.
* About generalizing `dataIF_Parallel`, I think these are our options, sorted by increasing difficulty:
   * Consider the current code a particular implementation limited to DC2, just like `DataIF_COSI_DC2`. In fact, this was one of the main reasons @hiyoneda worked out the interfaces to the imaging module, to decouple the algorithm from the data format. You can rename it ~`DataIF_Parallel`~ `DataIF_COSI_DC2_parallel` if you want.
   * You can also keep `DataIF_Parallel` as a parent class than handles everything except for the reading of the data, and then create a derived class  `DataIF_COSI_DC2_parallel(DataIF_Parallel)` that provides the I/O method.
   * Wait until the interfaces refactoring to work out with a generic abstract class.
   
Independently of which way you decide to go, I think this can be merged already and generalize it in a following PR. I added a note of this limitation (DC2 data and PSR only) to the description of the app.

avalluvan commented:
Hi Israel. I agree with these updates. We can keep the dataIF code as DC2-specific for now and expand it in a future PR. Glad this PR is close to being resolved.

israelmcmc commented:
Great! When you get a chance, can you please merge my PR to your branch, so this PR gets updated and all the tests are run?

avalluvan commented:
I already did. I am surprised the PR has not synced yet to something similar to this `https://github.com/cositools/cosipy/compare/develop...israelmcmc:PR274-Israel`

israelmcmc commented:
I'm closing and reopening the PR to kick the testing pipelines.

israelmcmc commented:
@avalluvan It seems all my commits and conflict resolutions were squashed into a single commit and, for reasons I don't fully understand without checking it in detail, this created some other conflict. This prevented all pipelines from running. 

I went ahead and opened #408 which has all of @avalluvan's and my commits, and doesn't have any conflicts. The pipelines run just fine. I'll close this PR and merge #408 instead.

@hiyoneda If you see any issue we can handle them in another PR. The non-MPI code was untouched, though, for the most part.
